src/java/org/apache/cassandra/nodes/virtual/LegacySystemKeyspaceToNodes.java

```diff
@@ -0,0 +1,286 @@
+
+/**
+ * Upgrade code to migrate from {@code system.local} and {@code system.peers} to {@link Nodes}.
+ */
+public final class LegacySystemKeyspaceToNodes
+{
+    private static final Logger logger = LoggerFactory.getLogger(LegacySystemKeyspaceToNodes.class);
+
+    private final Nodes nodes;
+
+    public int peersMigrated;
+    public boolean localMigrated;
+
+    public static boolean convertToNodesFilesIfNecessary()
+    {
+        if (Nodes.local().localInfoWasPresent())
+            // nothing to do
+            return false;
+
+        try
+        {
+            return new LegacySystemKeyspaceToNodes(Nodes.nodes()).convertToNodesFiles();
+        }
+        catch (Exception e)
+        {
+            throw new RuntimeException("Failed to migrate system.peers/local", e);
+        }
+    }
+
+    public LegacySystemKeyspaceToNodes(Nodes nodes)
+    {
+        this.nodes = nodes;
+    }
+
+    @VisibleForTesting
+    boolean convertToNodesFiles() throws Exception
+    {
+        Schema.unmapSystemTable(LOCAL);
+        Schema.unmapSystemTable(LEGACY_PEERS);
+        boolean localUpgraded = false;
+        try
+        {
+            Keyspace systemKs = Keyspace.open(SchemaConstants.SYSTEM_KEYSPACE_NAME);
+            ColumnFamilyStore cfs = systemKs.getColumnFamilyStore(SystemKeyspace.LEGACY_PEERS);
+            cfs.forceBlockingFlush(ColumnFamilyStore.FlushReason.STARTUP);
+            if (!cfs.getLiveSSTables().isEmpty())
+            {
+                logger.info("Migrating system.peers ...");
+
+                UntypedResultSet rows = executeInternal(format("SELECT * FROM %s.%s", SchemaConstants.SYSTEM_KEYSPACE_NAME, SystemKeyspace.LEGACY_PEERS));
+                if (!rows.isEmpty())
+                    migratePeers(nodes, rows);
+                else
+                    logger.info("Not migrated anything from system.peers because the table is empty...");
+
+                nodes.syncToDisk();
+
+                logger.info("{}.{} has been migrated and truncated.", SchemaConstants.SYSTEM_KEYSPACE_NAME, SystemKeyspace.LEGACY_PEERS);
+            }
+
+            cfs = systemKs.getColumnFamilyStore(SystemKeyspace.LOCAL);
+            cfs.forceBlockingFlush(ColumnFamilyStore.FlushReason.STARTUP);
+            if (!cfs.getLiveSSTables().isEmpty())
+            {
+                logger.info("Migrating system.local ...");
+
+                UntypedResultSet rows = executeInternal(format("SELECT * FROM %s.%s WHERE key = '%s'", SchemaConstants.SYSTEM_KEYSPACE_NAME, SystemKeyspace.LOCAL, SystemKeyspace.LOCAL));
+                if (!rows.isEmpty())
+                    migrateLocal(nodes, rows);
+                else
+                    logger.info("Not migrated anything from system.local because the table is empty...");
+
+                nodes.syncToDisk();
+
+                logger.info("{}.{} has been migrated and truncated.", SchemaConstants.SYSTEM_KEYSPACE_NAME, SystemKeyspace.LOCAL);
+
+                localUpgraded = true;
```

Comment:
Isn't it a bit fragile how you set this flag? only inside the if. Does it make sense there are no system.local to migrate and then you'd return false? It's bit confusing to read imo.

Reply:
This code is largely ported from DSE 6.8 and it seems that detail slipped my attention. The intention is to return `true` when anything is migrated, I will correct it as it seems to contain a bug - migrated peers return `false`. I will also add a java doc explaining the return value.

Reply:
In the end I have simplified it by removing the returned bool value - it didn't provide enough of a value.